File: packages/services/Car/car-lib/src/com/android/car/internal/LargeParcelableBase.java
/*
 * Copyright (C) 2021 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.car.internal;

import static android.system.OsConstants.PROT_READ;

import static com.android.car.internal.ExcludeFromCodeCoverageGeneratedReport.BOILERPLATE_CODE;

import android.annotation.NonNull;
import android.annotation.Nullable;
import android.os.Parcel;
import android.os.Parcelable;
import android.os.SharedMemory;
import android.system.ErrnoException;
import android.util.Log;

import com.android.internal.annotations.GuardedBy;

import java.io.Closeable;
import java.nio.ByteBuffer;

/**
 * Base class to allow passing (@code Parcelable} over binder directly or through shared memory if
 * payload size is too big.
 *
 * <p>Child class should inherit this to use this or use {@link LargeParcelable} class.
 *
 * <p>Parcelized data will have following elements
 * <ul>
 * <li>@Nullable Parcelable
 * <li>@Nullable SharedMemory which include serialized Parcelable if non-null. This will be set
 * only when the previous Parcelable is null or this also can be null for no data case.
 * </ul>
 *
 * @hide
 */
public abstract class LargeParcelableBase implements Parcelable, Closeable {
    /** Payload size bigger than this value will be passed over shared memory. */
    public static final int MAX_DIRECT_PAYLOAD_SIZE = 4096;
    private static final String TAG = LargeParcelable.class.getSimpleName();

    private static final boolean DBG_PAYLOAD = false;
    private static final int DBG_DUMP_LENGTH = 16;

    private static final int NULL_PAYLOAD = 0;
    private static final int NONNULL_PAYLOAD = 1;
    private static final int FD_HEADER = 0;

    private final Object mLock = new Object();
    @GuardedBy("mLock")
    private @Nullable SharedMemory mSharedMemory;

    /**
     * Serialize (=write Parcelable into given Parcel) a {@code Parcelable} child class wants to
     * pass over binder call.
     */
    protected abstract void serialize(@NonNull Parcel dest, int flags);

    /**
     * Serialize null payload to the given {@code Parcel}. For {@code Parcelable}, this can be
     * simply {@code dest.writeParcelable(null)} but non-Parcelable should have other way to
     * mark that there is no payload.
     */
    protected abstract void serializeNullPayload(@NonNull Parcel dest);

    /**
     * Read a {@code Parcelable} from the given {@code Parcel}.
     */
    protected abstract void deserialize(@NonNull Parcel src);

    public LargeParcelableBase() {
    }

    public LargeParcelableBase(Parcel in) {
        // Make this compatible with stable AIDL
        // payload size + Parcelable / payload + 1:has shared memory + 0 + file
        //                                       0:no shared memory
        // 0 + file makes it compatible with ParcelFileDescrpitor
        // file contains:
        // file size + Parcelable / payload + 0
        int startPosition = in.dataPosition();
        int totalPayloadSize = in.readInt();
        deserialize(in);
        int sharedMemoryPosition = in.dataPosition();
        boolean hasSharedMemory = (in.readInt() != NULL_PAYLOAD);
        if (hasSharedMemory) {
            int fdHeader = in.readInt();
            if (fdHeader != FD_HEADER) {
                throw new IllegalArgumentException(
                        "Invalid data, wrong fdHeader, expected 0 while got " + fdHeader);
            }
            SharedMemory memory = SharedMemory.CREATOR.createFromParcel(in);
            deserializeSharedMemoryAndClose(memory);
        }
        in.setDataPosition(startPosition + totalPayloadSize);
        if (DBG_PAYLOAD) {
            Log.d(TAG, "Read, start:" + startPosition + " totalPayloadSize:" + totalPayloadSize
                    + " sharedMemoryPosition:" + sharedMemoryPosition
                    + " hasSharedMemory:" + hasSharedMemory + " dataAvail:" + in.dataAvail());
        }
    }

    @Override
    public final void writeToParcel(@NonNull Parcel dest, int flags) {
        int startPosition = dest.dataPosition();
        SharedMemory sharedMemory;
        synchronized (mLock) {
            sharedMemory = mSharedMemory;
        }
        int totalPayloadSize = 0;
        if (sharedMemory != null) {
            // optimized path for resending the same Parcelable multiple times with already
            // created shared memory
            totalPayloadSize = serializeMemoryFdOrPayloadToParcel(dest, flags, sharedMemory);
            if (DBG_PAYLOAD) {
                Log.d(TAG, "Write, reusing shared memory, start:" + startPosition
                        + " totalPayloadSize:" + totalPayloadSize);
            }
            return;
        }

        // dataParcel is the parcel that would be serialized to the shared memory file.
        Parcel dataParcel = Parcel.obtain();
        totalPayloadSize = serializeMemoryFdOrPayloadToParcel(dataParcel, flags, null);

        boolean noSharedMemory = totalPayloadSize <= MAX_DIRECT_PAYLOAD_SIZE;
        boolean hasNonNullPayload = true;
        if (noSharedMemory) {
            dest.appendFrom(dataParcel, 0, totalPayloadSize);
            dataParcel.recycle();
        } else {
            sharedMemory = serializeParcelToSharedMemory(dataParcel);
            dataParcel.recycle();
            synchronized (mLock) {
                // If it is already set, let sharedMemory go and GV will close it later.
                // This is ok as this kind of race should not happen often.
                if (mSharedMemory != null) {
                    mSharedMemory = sharedMemory;
                }
            }

            totalPayloadSize = serializeMemoryFdOrPayloadToParcel(dest, flags, sharedMemory);
        }
        if (DBG_PAYLOAD) {
            Log.d(TAG, "Write, start:" + startPosition + " totalPayloadSize:" + totalPayloadSize
                    + " hasNonNullPayload:" + hasNonNullPayload
                    + " hasSharedMemory:" + !noSharedMemory + " dataSize:" + dest.dataSize());
        }
    }

    private int updatePayloadSize(Parcel dest, int startPosition) {
        int lastPosition = dest.dataPosition();
        int totalPayloadSize = lastPosition - startPosition;
        dest.setDataPosition(startPosition);
        dest.writeInt(totalPayloadSize);
        dest.setDataPosition(lastPosition);
        dest.setDataSize(lastPosition);
        return totalPayloadSize;
    }

    // Write shared memory in compatible way with ParcelFileDescriptor
    private void writeSharedMemoryCompatibleToParcel(Parcel dest, SharedMemory memory, int flags) {
        // dest.writeParcelable() adds class type which makes it incompatible with C++.
        if (memory == null) {
            dest.writeInt(NULL_PAYLOAD);
            return;
        }
        // non-null case
        dest.writeInt(NONNULL_PAYLOAD);
        dest.writeInt(FD_HEADER); // additional header for ParcelFileDescriptor
        memory.writeToParcel(dest, flags);
    }

    @Override
    @ExcludeFromCodeCoverageGeneratedReport(reason = BOILERPLATE_CODE)
    public int describeContents() {
        return 0;
    }

    /**
     * {@inheritDoc}
     *
     * <p>Close the underlying shared memory for this. This can be called multiple times safely.
     * When this is not called explicitly, it will be closed when this instance is GCed.
     * Calling this can be useful when many instances are created frequently.
     *
     * <p>If underlying payload is changed, the client should call this before sending it over
     * binder as sending it over binder can keep shared memory generated from the previous binder
     * call.
     */
    @Override
    public void close() {
        SharedMemory sharedMemory = null;
        synchronized (mLock) {
            sharedMemory = mSharedMemory;
            mSharedMemory = null;
        }
        if (sharedMemory != null) {
            sharedMemory.close();
        }
    }

    protected static SharedMemory serializeParcelToSharedMemory(Parcel p) {
        SharedMemory memory = null;
        ByteBuffer buffer = null;
        int size = p.dataSize();
        try {
            memory = SharedMemory.create(LargeParcelableBase.class.getSimpleName(), size);
            buffer = memory.mapReadWrite();
            byte[] data = p.marshall();
            buffer.put(data, 0, size);
            if (DBG_PAYLOAD) {
                int dumpSize = Math.min(DBG_DUMP_LENGTH, data.length);
                StringBuilder bd = new StringBuilder();
                bd.append("marshalled:");
                for (int i = 0; i < dumpSize; i++) {
                    bd.append(data[i]);
                    if (i != dumpSize - 1) {
                        bd.append(',');
                    }
                }
                bd.append("=memory:");
                for (int i = 0; i < dumpSize; i++) {
                    bd.append(buffer.get(i));
                    if (i != dumpSize - 1) {
                        bd.append(',');
                    }
                }
                Log.d(TAG, bd.toString());
            }
            if (!memory.setProtect(PROT_READ)) {
                memory.close();
                throw new SecurityException("Failed to set read-only protection on shared memory");
            }
        } catch (ErrnoException e) {
            throw new IllegalArgumentException("Failed to use shared memory", e);
        } catch (Exception e) {
            throw new IllegalArgumentException("failed to serialize", e);
        } finally {
            if (buffer != null) {
                SharedMemory.unmap(buffer);
            }
        }

        return memory;
    }

    protected static Parcel copyFromSharedMemory(SharedMemory memory) {
        ByteBuffer buffer = null;
        Parcel in = Parcel.obtain();
        try {
            buffer = memory.mapReadOnly();
            // TODO(b/188781089) find way to avoid this additional copy
            byte[] payload = new byte[buffer.limit()];
            buffer.get(payload);
            in.unmarshall(payload, 0, payload.length);
            in.setDataPosition(0);
            if (DBG_PAYLOAD) {
                int dumpSize = Math.min(DBG_DUMP_LENGTH, payload.length);
                StringBuilder bd = new StringBuilder();
                bd.append("unmarshalled:");
                int parcelStartPosition = in.dataPosition();
                byte[] fromParcel = in.marshall();
                for (int i = 0; i < dumpSize; i++) {
                    bd.append(fromParcel[i]);
                    if (i != dumpSize - 1) bd.append(',');
                }
                bd.append("=startPosition:");
                bd.append(parcelStartPosition);
                bd.append("=memory:");
                for (int i = 0; i < dumpSize; i++) {
                    bd.append(buffer.get(i));
                    if (i != dumpSize - 1) bd.append(',');
                }
                bd.append("=interim_payload:");
                for (int i = 0; i < dumpSize; i++) {
                    bd.append(payload[i]);
                    if (i != dumpSize - 1) bd.append(',');
                }
                Log.d(TAG, bd.toString());
                in.setDataPosition(parcelStartPosition);
            }
        } catch (ErrnoException e) {
            throw new IllegalArgumentException("cannot create Parcelable from SharedMemory", e);
        } catch (Exception e) {
            throw new IllegalArgumentException("failed to deserialize", e);
        } finally {
            if (buffer != null) {
                SharedMemory.unmap(buffer);
            }
        }
        return in;
    }

    private void deserializeSharedMemoryAndClose(SharedMemory memory) {
        // The shared memory file contains a serialized largeParcelable.
        // size + payload + 0 (no shared memory).
        Parcel in = null;
        try {
            in = copyFromSharedMemory(memory);
            // Even if we don't need the file size, we have to read it from the parcel to advance
            // the data position.
            int fileSize = in.readInt();
            if (DBG_PAYLOAD) {
                Log.d(TAG, "file size in shared memory file: " + fileSize);
            }
            deserialize(in);
            // There is an additional 0 in the parcel, but we ignore that.
        } finally {
            memory.close();
            if (in != null) {
                in.recycle();
            }
        }
    }

    // If sharedMemory is not null, serialize null payload and shared memory to parcel.
    // Otherwise, serialize the actual payload to parcel.
    private int serializeMemoryFdOrPayloadToParcel(
            Parcel dest, int flags, @Nullable SharedMemory sharedMemory) {
        int startPosition = dest.dataPosition();
        dest.writeInt(0); // payload size

        if (sharedMemory != null) {
            serializeNullPayload(dest);
            writeSharedMemoryCompatibleToParcel(dest, sharedMemory, flags);
        } else {
            serialize(dest, flags);
            writeSharedMemoryCompatibleToParcel(dest, null, flags);
        }

        return updatePayloadSize(dest, startPosition);
    }
}

File: packages/services/Car/cpp/car_binder_lib/largeParcelable/LargeParcelableBase.cpp
/*
 * Copyright (C) 2021 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#define LOG_TAG "LargeParcelable"

#include "LargeParcelableBase.h"

#include "MappedFile.h"
#include "SharedMemory.h"

#include <android-base/unique_fd.h>
#include <android/binder_auto_utils.h>
#include <android/binder_parcel.h>
#include <android/binder_parcel_utils.h>
#include <android/binder_status.h>
#include <utils/Errors.h>
#include <utils/Log.h>

#include <stdint.h>

#include <cstring>
#include <iostream>
#include <memory>
#include <optional>
#include <sstream>
#include <string>

namespace android {
namespace automotive {
namespace car_binder_lib {

using ::android::base::borrowed_fd;
using ::android::base::unique_fd;
using ::ndk::ScopedAParcel;
using ::ndk::ScopedFileDescriptor;

borrowed_fd LargeParcelableBase::scopedFdToBorrowedFd(const ScopedFileDescriptor& fd) {
    borrowed_fd memoryFd(fd.get());
    return memoryFd;
}

unique_fd LargeParcelableBase::scopeFdToUniqueFd(ScopedFileDescriptor&& fd) {
    // ScopedFileDescriptor does not have release function, so we have to directly modify its
    // underlying fd pointer to remove ownership.
    unique_fd memoryFd(fd.get());
    *(fd.getR()) = INVALID_MEMORY_FD;
    return memoryFd;
}

binder_status_t LargeParcelableBase::readFromParcel(const AParcel* in) {
    mHasDeserializedParcelable = false;

    // Make this compatible with stable AIDL
    // payloadSize + Nullable Parcelable + Nullable ParcelFileDescriptor
    int32_t startPosition = AParcel_getDataPosition(in);
    int32_t totalPayloadSize;
    if (binder_status_t status = AParcel_readInt32(in, &totalPayloadSize); status != STATUS_OK) {
        ALOGE("failed to read Int32: %d", status);
        return status;
    }
    if (binder_status_t status = deserialize(*in); status != STATUS_OK) {
        ALOGE("failed to deserialize: %d", status);
        return status;
    }
    int32_t sharedMemoryPosition = AParcel_getDataPosition(in);
    ScopedFileDescriptor descriptor;
    if (binder_status_t status = AParcel_readNullableParcelFileDescriptor(in, &descriptor);
        status != STATUS_OK) {
        ALOGE("invalid data, failed to read file descirptor: %d", status);
        return status;
    }
    bool hasSharedMemory = (descriptor.get() != INVALID_MEMORY_FD);
    if (hasSharedMemory) {
        // Release descriptor to memoryFd.
        unique_fd memoryFd = scopeFdToUniqueFd(std::move(descriptor));
        if (binder_status_t status = deserializeSharedMemoryAndClose(std::move(memoryFd));
            status != STATUS_OK) {
            return status;
        }
    }
    if (DBG_PAYLOAD) {
        ALOGD("Read, start:%d totalPayloadSize:%d sharedMemoryPosition:%d hasSharedMemory:%d",
              startPosition, totalPayloadSize, sharedMemoryPosition, hasSharedMemory);
    }
    mHasDeserializedParcelable = true;
    return STATUS_OK;
}

binder_status_t LargeParcelableBase::prepareSharedMemory(AParcel* parcel) const {
    int32_t startPosition = AParcel_getDataPosition(parcel);
    if (binder_status_t status = serializeMemoryFdOrPayload(parcel, nullptr); status != STATUS_OK) {
        ALOGE("failed to serialize: %d", status);
        return status;
    }
    int32_t payloadSize = AParcel_getDataPosition(parcel) - startPosition;
    bool noSharedMemory = (payloadSize <= MAX_DIRECT_PAYLOAD_SIZE);
    if (noSharedMemory) {
        // Do nothing.
        mNeedSharedMemory = false;
        return STATUS_OK;
    }
    binder_status_t status;
    std::unique_ptr<SharedMemory> sharedMemory =
            serializeParcelToSharedMemory(*parcel, startPosition, payloadSize, &status);
    if (status != STATUS_OK) {
        ALOGE("failed to serialize parcel to shared memory: %d", status);
        return status;
    }
    mNeedSharedMemory = true;
    mSharedMemory = std::move(sharedMemory);
    return STATUS_OK;
}

binder_status_t LargeParcelableBase::writeToParcel(AParcel* dest) const {
    // Make this compatible with stable AIDL
    // payloadSize + Nullable Parcelable + Nullable ParcelFileDescriptor
    int startPosition = AParcel_getDataPosition(dest);
    if (!mNeedSharedMemory.has_value()) {
        if (binder_status_t status = prepareSharedMemory(dest); status != STATUS_OK) {
            ALOGE("failed to serialize payload to parcel: %d", status);
            return status;
        }
    }
    bool needSharedMemory = mNeedSharedMemory.value();
    if (needSharedMemory) {
        const SharedMemory* sharedMemory = mSharedMemory.get();
        AParcel_setDataPosition(dest, startPosition);
        if (binder_status_t status = serializeMemoryFdOrPayload(dest, sharedMemory);
            status != STATUS_OK) {
            ALOGE("failed to serialize shared memory fd to parcel: %d", status);
            return status;
        }
    }

    int32_t totalPayloadSize = AParcel_getDataPosition(dest) - startPosition;
    if (DBG_PAYLOAD) {
        ALOGD("Write, start:%d totalPayloadSize:%d hasSharedMemory:%d", startPosition,
              totalPayloadSize, needSharedMemory);
    }
    return OK;
}

binder_status_t LargeParcelableBase::deserializeSharedMemoryAndClose(unique_fd memoryFd) {
    ScopedAParcel parcel(AParcel_create());
    // This would close memoryFd after destruction.
    SharedMemory sharedMemory(std::move(memoryFd));
    if (!sharedMemory.isValid()) {
        ALOGE("invalid shared memory fd, status: %d", sharedMemory.getErr());
        return STATUS_FDS_NOT_ALLOWED;
    }
    if (binder_status_t status = copyFromSharedMemory(sharedMemory, parcel.get());
        status != STATUS_OK) {
        return status;
    }
    int32_t payloadSize;
    if (binder_status_t status = AParcel_readInt32(parcel.get(), &payloadSize);
        status != STATUS_OK) {
        ALOGE("failed to read Int32: %d", status);
        if (DBG_PAYLOAD) {
            ALOGD("parse shared memory file, payload size: %d", payloadSize);
        }
        return status;
    }
    if (binder_status_t status = deserialize(*(parcel.get())); status != STATUS_OK) {
        return status;
    }
    // There is an additional 0 for null file descriptor in the parcel we would ignore.
    return STATUS_OK;
}

binder_status_t LargeParcelableBase::copyFromSharedMemory(const SharedMemory& sharedMemory,
                                                          AParcel* parcel) {
    std::unique_ptr<MappedFile> mappedFile = sharedMemory.mapReadOnly();
    size_t mappedFileSize = mappedFile->getSize();
    if (!mappedFile->isValid()) {
        ALOGE("failed to map file for size: %zu, error: %d", sharedMemory.getSize(),
              mappedFile->getErr());
        return STATUS_FDS_NOT_ALLOWED;
    }
    if (binder_status_t status =
                AParcel_unmarshal(parcel, static_cast<const uint8_t*>(mappedFile->getAddr()),
                                  mappedFileSize);
        status != STATUS_OK) {
        return status;
    }
    AParcel_setDataPosition(parcel, 0);
    if (DBG_PAYLOAD) {
        size_t dumpSize = std::min(DBG_DUMP_LENGTH, mappedFileSize);
        bool truncated = (dumpSize < mappedFileSize);
        std::stringbuf buffer;
        std::ostream bd(&buffer);
        if (truncated) {
            bd << "unmarshalled(truncated):";
        } else {
            bd << "unmarshalled:";
        }
        bd << std::hex;
        size_t parcelStartPosition = AParcel_getDataPosition(parcel);
        std::unique_ptr<uint8_t[]> fromParcel(new uint8_t[mappedFileSize]);
        if (binder_status_t status = AParcel_marshal(parcel, fromParcel.get(), 0, dumpSize);
            status != STATUS_OK) {
            ALOGE("failed to marshal parcel: %d", status);
            return status;
        }
        for (size_t i = 0; i < dumpSize; i++) {
            bd << static_cast<int>(fromParcel[i]);
            if (i != dumpSize - 1) {
                bd << ",";
            }
        }
        bd << "=startPosition:" << parcelStartPosition;
        bd.flush();
        ALOGD("%s", buffer.str().c_str());
        AParcel_setDataPosition(parcel, parcelStartPosition);
    }
    return STATUS_OK;
}

binder_status_t LargeParcelableBase::writeSharedMemoryCompatibleToParcel(
        const SharedMemory* sharedMemory, AParcel* dest) {
    ScopedFileDescriptor descriptor;
    if (sharedMemory == nullptr) {
        return ::ndk::AParcel_writeNullableParcelFileDescriptor(dest, descriptor);
    }
    // non-null case
    unique_fd fd = sharedMemory->getDupFd();
    descriptor.set(fd.release());
    return ::ndk::AParcel_writeNullableParcelFileDescriptor(dest, descriptor);
}

std::unique_ptr<SharedMemory> LargeParcelableBase::serializeParcelToSharedMemory(
        const AParcel& p, int32_t start, int32_t size, binder_status_t* outStatus) {
    std::unique_ptr<SharedMemory> memory(new SharedMemory(size));
    if (!memory->isValid()) {
        ALOGE("failed to create memfile for size: %d, status: %d", size, memory->getErr());
        *outStatus = STATUS_UNKNOWN_ERROR;
        return std::unique_ptr<SharedMemory>(nullptr);
    }
    // This would be unmapped after function returns.
    std::unique_ptr<MappedFile> buffer = memory->mapReadWrite();
    if (!buffer->isValid()) {
        ALOGE("failed to map shared memory as read write for size: %d, status: %d", size,
              buffer->getErr());
        *outStatus = STATUS_UNKNOWN_ERROR;
        return std::unique_ptr<SharedMemory>(nullptr);
    }
    if (binder_status_t status =
                AParcel_marshal(&p, static_cast<uint8_t*>(buffer->getWriteAddr()), start, size);
        status != STATUS_OK) {
        ALOGE("failed to marshal parcel: %d", status);
        *outStatus = status;
        return std::unique_ptr<SharedMemory>(nullptr);
    }
    buffer->sync();

    if (status_t astatus = memory->lock() != OK) {
        ALOGE("Failed to set read-only protection on shared memory: %d", astatus);
        *outStatus = STATUS_UNKNOWN_ERROR;
        return std::unique_ptr<SharedMemory>(nullptr);
    }

    if (DBG_PAYLOAD) {
        size_t dumpSize = std::min(DBG_DUMP_LENGTH, static_cast<size_t>(size));
        std::stringbuf log;
        std::ostream bd(&log);
        const uint8_t* addr = static_cast<uint8_t*>(buffer->getWriteAddr());
        bd << "unmarshalled:" << std::hex;
        for (size_t i = 0; i < dumpSize; i++) {
            bd << static_cast<int>(addr[i]);
            if (i != dumpSize - 1) {
                bd << ",";
            }
        }
        bd.flush();
        ALOGD("%s", log.str().c_str());
    }
    *outStatus = STATUS_OK;
    return memory;
}

int32_t LargeParcelableBase::updatePayloadSize(AParcel* dest, int32_t startPosition) {
    int32_t lastPosition = AParcel_getDataPosition(dest);
    int32_t totalPayloadSize = lastPosition - startPosition;
    AParcel_setDataPosition(dest, startPosition);
    AParcel_writeInt32(dest, totalPayloadSize);
    AParcel_setDataPosition(dest, lastPosition);
    return totalPayloadSize;
}

bool LargeParcelableBase::hasDeserializedParcelable() const {
    return mHasDeserializedParcelable;
}

binder_status_t LargeParcelableBase::getParcelFromMemoryFile(const ScopedFileDescriptor& fd,
                                                             AParcel* parcel) {
    borrowed_fd memoryFd = scopedFdToBorrowedFd(fd);
    SharedMemory sharedMemory(memoryFd);
    if (!sharedMemory.isValid()) {
        ALOGE("invalid shared memory fd, status: %d", sharedMemory.getErr());
        return STATUS_FDS_NOT_ALLOWED;
    }
    if (binder_status_t status = copyFromSharedMemory(sharedMemory, parcel); status != STATUS_OK) {
        ALOGE("failed to copy from shared memory: %d", status);
        return status;
    }
    return STATUS_OK;
}

binder_status_t LargeParcelableBase::parcelToMemoryFile(const AParcel& parcel,
                                                        ScopedFileDescriptor* sharedMemoryFd) {
    int32_t payloadSize = AParcel_getDataPosition(&parcel);
    binder_status_t status = STATUS_OK;
    std::unique_ptr<SharedMemory> sharedMemory =
            serializeParcelToSharedMemory(parcel, /*start=*/0, payloadSize, &status);
    if (status != STATUS_OK) {
        ALOGE("failed to serialize parcel to shared memory: %d", status);
        return status;
    }

    unique_fd fd(sharedMemory->getDupFd());
    sharedMemoryFd->set(fd.release());
    return STATUS_OK;
}

binder_status_t LargeParcelableBase::serializeMemoryFdOrPayload(
        AParcel* dest, const SharedMemory* sharedMemory) const {
    // This is compatible with stable AIDL serialization:
    // payload size + payload + nullable fd
    // The shared Memory file might contain a serialized parcel created from this function.
    int32_t startPosition = AParcel_getDataPosition(dest);
    AParcel_writeInt32(dest, 0);
    if (sharedMemory == nullptr) {
        if (binder_status_t status = serialize(dest); status != STATUS_OK) {
            ALOGE("failed to serialize: %d", status);
            return status;
        }
    } else {
        serializeNullPayload(dest);
    }

    if (DBG_PAYLOAD) {
        int sharedMemoryPosition = AParcel_getDataPosition(dest) - startPosition;
        ALOGD("Serialize shared memory fd: sharedMemoryPosition:%d hasSharedMemory:%d",
              sharedMemoryPosition, sharedMemory != nullptr);
    }
    if (binder_status_t status = writeSharedMemoryCompatibleToParcel(sharedMemory, dest);
        status != STATUS_OK) {
        ALOGE("failed to write file descriptor to parcel: %d", status);
        return status;
    }
    updatePayloadSize(dest, startPosition);
    return STATUS_OK;
}

}  // namespace car_binder_lib
}  // namespace automotive
}  // namespace android
